rustls/src/msgs/deframer/handshake.rs

```diff
@@ -0,0 +1,543 @@
+use alloc::vec::Vec;
+use core::mem;
+use core::ops::Range;
+
+use super::buffers::{Coalescer, Delocator, Locator};
+use crate::msgs::codec::{u24, Codec};
+use crate::msgs::message::InboundPlainMessage;
+use crate::{ContentType, ProtocolVersion};
+
+#[derive(Debug)]
+pub(crate) struct HandshakeDeframer {
+    /// Spans covering individual handshake payloads, in order of receipt.
+    spans: Vec<FragmentSpan>,
+}
+
+impl HandshakeDeframer {
+    /// Accepts a message into the deframer.
+    ///
+    /// `containing_buffer` allows mapping the message payload to its position
+    /// in the input buffer, and thereby avoid retaining a borrow on the input
+    /// buffer.
+    ///
+    /// That is required because our processing of handshake messages requires
+    /// them to be contiguous (and avoiding that would mean supporting gather-based
+    /// parsing in a large number of places, including `core`, `webpki`, and the
+    /// `CryptoProvider` interface).  `coalesce()` arranges for that to happen, but
+    /// to do so it needs to move the fragments together in the original buffer.
+    /// This would not be possible if the messages were borrowing from that buffer.
+    ///
+    /// `outer_discard` is the rightmost extent of the original message.
+    pub(crate) fn input_message(
+        &mut self,
+        msg: InboundPlainMessage<'_>,
+        containing_buffer: &Locator,
+        outer_discard: usize,
+    ) {
+        debug_assert_eq!(msg.typ, ContentType::Handshake);
+        debug_assert!(containing_buffer.fully_contains(msg.payload));
+
+        // if our last span is incomplete, we can blindly add this as a new span --
+        // no need to attempt parsing it with `DissectHandshakeIter`.
+        //
+        // `coalesce()` will later move this new message to be contiguous with
+        // `_last_incomplete`, and reparse the result.
+        //
+        // we cannot merge these processes, because `coalesce` mutates the underlying
+        // buffer, and `msg` borrows it.
```

Comment:
I view `InboundUnborrowedMessage` as a bit of a hack as part of convincing the borrow checker to determine the correct lifetime relationship between the buffer and object returned by `deframe()`. So while this would work equally well, perhaps better, I'm minded to keep it as is.